### PR TITLE
Add the `InterfaceName` option to `Pinger`

### PR DESCRIPTION
**Pull request by MatthieuCoder**
This is a rebase of #32 to match the current HEAD on main.

This allows using `pro-bing` to use VRFs interfaces and IPv6 link-local addresses.

### CODE REVIEW

File: ping.go [hunk 1]
```diff
@@ -207,6 +207,9 @@ type Pinger struct {
 	// Source is the source IP address
 	Source string
 
+	// Interface used to send/recv ICMP messages
+	Interface string
```
Comment (SuperQ) [inline ping.go:211]: I think we wanted to have this be something `InterfaceDevice` to avoid conflict with the `interface` keyword.

Comment (MatthieuCoder) [reply ping.go]: I changed it to `InterfaceDevice`, but wouldn't `InterfaceName` be more appropriate ?

Comment (SuperQ) [reply ping.go]: I don't have a strong opinion here. `InterfaceName` also works for me. Up to you.

**Comment by SuperQ** [comment]
It seems like the test fails on permissions.

**Comment by MatthieuCoder** [comment]
As @ilolicon pointed out, the tests run just fine on a local hosts & using the circleci local cli:

```
matthieu@matthieu-archlinux ~/pro-bing (main)> circleci local execute build
Fetching latest build environment...
Docker image digest: sha256:008ba7f4223f1e26c11df9575283491b620074fa96da6961e0dcde47fb757014
====>> Spin up environment
Build-agent version  ()
System information:
 Server Version: 27.3.1
 Storage Driver: overlay2
  Backing Filesystem: btrfs
 Cgroup Driver: systemd
 Cgroup Version: 2
 Kernel Version: 6.11.5-arch1-1
 Operating System: Arch Linux
 OSType: linux
 Architecture: x86_64

Starting container cimg/go:1.20
Warning: No authentication provided, using CircleCI credentials for pulls from Docker Hub.
  image is cached as cimg/go:1.20, but refreshing...
1.20: Pulling from cimg/go
Digest: sha256:bd1abc34d5deb0c8f55e9f9c8071210e3a5ba9ce874f48a61a390175c2be0da6
Status: Image is up to date for cimg/go:1.20
cimg/go:1.20:
  using image cimg/go@sha256:bd1abc34d5deb0c8f55e9f9c8071210e3a5ba9ce874f48a61a390175c2be0da6
  pull stats: Image was already available so the image was not pulled
  time to create container: 117ms
Creating Docker containers in parallel 
Time to upload agent and config: 162.595656ms
Time to start containers: 457.684773ms
====>> Preparing environment variables
Using build environment variables:
  BASH_ENV=/tmp/.bash_env-localbuild-1730205231
  CI=true
  CIRCLECI=true
  CIRCLE_BRANCH=
  CIRCLE_BUILD_NUM=
  CIRCLE_JOB=build
  CIRCLE_NODE_INDEX=0
  CIRCLE_NODE_TOTAL=1
  CIRCLE_REPOSITORY_URL=
  CIRCLE_SHA1=
  CIRCLE_SHELL_ENV=/tmp/.bash_env-localbuild-1730205231
  CIRCLE_WORKING_DIRECTORY=~/project


The redacted variables listed above will be masked in run step output.====>> Checkout code
Making checkout directory "/home/circleci/project"
Copying files from "/tmp/_circleci_local_build_repo" to "/home/circleci/project"
====>> go mod download
  #!/bin/bash -eo pipefail
go mod download
====>> make
  #!/bin/bash -eo pipefail
make
>> checking code style
>> vetting code
GO111MODULE= go vet  ./...
>> building ping
GO111MODULE= go build  ./cmd/ping
>> running all tests
GO111MODULE= go test -race -cover  ./...
?       github.com/prometheus-community/pro-bing/cmd/ping       [no test files]
ok      github.com/prometheus-community/pro-bing        7.105s  coverage: 67.9% of statements
Success!
```

**Comment by MatthieuCoder** [comment]
Seems like using another interface did not fix the issue regarding permissions,
Specifying an interface might be unsupported in CircleCI ?

**Comment by SuperQ** [comment]
We can try switching from Docker to Machine mode for CI.

**Comment by MatthieuCoder** [comment]
I would like that.

**Comment by MatthieuCoder** [comment]
Do I need to pull the changes on `main` for them to apply here ?

**Comment by MatthieuCoder** [comment]
Permission denied still, I'll look into it tomorrow

**Comment by MatthieuCoder** [comment]
As I thought, Using the machine executor does make the tests pass,
Subsequently the change from docker to machine should be in another PR, since @SuperQ's PR does not (#124) does not change the execution to the machine type

**Comment by SuperQ** [comment]
Yea, the only problem with the machine image is it doesn't directly control the Go version.

**Comment by MatthieuCoder** [comment]
> Yea, the only problem with the machine image is it doesn't directly control the Go version.

Indeed, I read online that this could be solved by running docker in the machine executor, this might require more configuration.

**Comment by MatthieuCoder** [comment]
hold on I have a last thing to change

**Comment by MatthieuCoder** [comment]
now ready!

File: ping.go (hunk 1 above)
Comment (Hipska) [reply ping.go]: I would also pick `InterfaceName`.

File: ping.go [hunk 2]
```diff
@@ -207,6 +207,9 @@ type Pinger struct {
 	// Source is the source IP address
 	Source string
 
+	// Interface used to send/recv ICMP messages
+	InterfaceDevice string
```
Comment (Hipska) [inline ping.go:211]: Please use `InterfaceName`, it will make more sense with the `net.InterfaceByName` call 😉

**Review by Hipska — CHANGES_REQUESTED** [review]

**Review by Hipska — APPROVED** [review]

**Review by SuperQ — APPROVED** [review]
Thanks. We can fix the Go version matrix later. For now we can live with the version provide in the machine image.

**Comment by dswarbrick** [comment]
Tests are failing locally for me, on Ubuntu 24.10, running the test as root.
```
--- FAIL: TestSetInterfaceName (0.00s)
    ping_test.go:490: Expected No Error but got socket: permission denied, Stack:
        goroutine 259 [running]:
        runtime/debug.Stack()
                /usr/lib/go-1.23/src/runtime/debug/stack.go:26 +0x5e
        github.com/prometheus-community/pro-bing.AssertNoError(0xc0003844e0, {0x871880, 0xc000098260})
                /home/daniel/Work/Debian/golang-github-prometheus-community-pro-bing/ping_test.go:528 +0x37
        github.com/prometheus-community/pro-bing.TestSetInterfaceName(0xc0003844e0)
                /home/daniel/Work/Debian/golang-github-prometheus-community-pro-bing/ping_test.go:490 +0x95
        testing.tRunner(0xc0003844e0, 0x80d9a0)
                /usr/lib/go-1.23/src/testing/testing.go:1690 +0xf4
        created by testing.(*T).Run in goroutine 1
                /usr/lib/go-1.23/src/testing/testing.go:1743 +0x390
```

**Pull request merged**